src/inmanta/util.py

```diff
@@ -200,31 +201,57 @@ def get_free_tcp_port() -> str:
         return str(port)
 
 
-def custom_json_encoder(o: object) -> Union[Dict, str, List]:
+class JSONSerializable(ABC):
     """
-        A custom json encoder that knows how to encode other types commonly used by Inmanta from standard python libraries
+        Instances of this class are JSON serializable. Concrete subclasses should implement json_serialization_step.
     """
-    if isinstance(o, uuid.UUID):
-        return str(o)
 
-    if isinstance(o, datetime.datetime):
-        return o.isoformat(timespec="microseconds")
+    @abstractmethod
+    def json_serialization_step(self) -> Union[ReturnTypes, "JSONSerializable"]:
+        """
+            Perform a step in the serialization process. Returns an other serializable object.
+            The implementation should make sure each step progresses serialization so that successively
+            calling JSONSerializable.default eventually resolves to a trivially serializable object.
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def default(obj: object) -> Union[ReturnTypes, "JSONSerializable"]:
```

Comment:
why move this into a method on JSONSerializable? 

(or more specifically, what is the performance impact of this change?)

Reply:
I don't think there's any performance impact. Also, I don't think there's a good reason to do so (I made my design while unaware of the existing method and blindly stuck with it). I'll restore the original method.